Fix confidentialTransferFrom() for ZkAssetMintable

thomas-waite opened the pull request:
**Note: This PR should not be merged into master unless automatic release to NPM is disabled.**

## Description
This PR solves a problem in that `confidentialTransferFrom()` for the `ZkAssetMintable.sol` contract did not mint tokens appropriately. 

If a user minted tokens using `confidentialMint()` and then converted/transferred them out using a joinSplit proof – it is possible that the tokens attempted to be transferred out are greater than the total number held by the contract, and so the call will revert.

To address this, logic has been added, similar to `confidentialTransfer()` that will mint the appropriate number of tokens from the linked ERC20 token to cover the shortfall. 

<!--- Describe your changes in detail -->

## Testing instructions
`npm run test`

I have added two success test cases to `ZkAssetMintable.js`. 

The first checks that when ACE has insufficient number of tokens, that the correct number are minted when `confidentialTransferFrom()` is called.

The second checks that ACE does not mint tokens unnecessarily when it has enough i.e. if publicValue <= totalSupply

<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->
Bug fix - `confidentialTransferFrom() in `ZkAssetMintable.sol` did not correctly mint tokens.
<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->
## Notes
I am aware I have introduced some WET code in the redefining of `confidentialTransferFrom() in `ZkAssetMintable.sol`. 

Originally, I was going to pursue the approach taken in redefining `confidentialTransfer()` for `ZkAssetMintable` by defining a `confidentialTransferFromInternal()` in the parent `ZkAsset`.  However, realised that the proposed `confidentialTransferFromInternal()` and `confidentialTransferInternal()` would have the same code/perform the same function - and both could be combined into a more descriptive function such as `processTransaction()`. 

Will leave for a future refactor and focus this PR solely on the bug fix.

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.

thomas-waite committed:
fix(protocol): make confidentialTransferFrom() mint/supplement tokens if needed

thomas-waite committed:
test(protocol): add a ZkAssetMintableTest contract to test confidentialTransferFrom()

thomas-waite committed:
test(protocol): success test to check mint in confidentialTransferFrom()

thomas-waite committed:
test(protocol): success test to check no minting when not required

thomas-waite committed:
docs(protocol): update docs for ZkAssetMintable functions

ArnSch commented on packages/protocol/test/ZkAsset/ZkAssetMintable.js:
```diff
@@ -15,13 +15,28 @@ const AdjustSupply = artifacts.require('./contracts/ACE/validators/AdjustSupply'
 AdjustSupply.abi = AdjustSupplyInterface.abi;
 JoinSplit.abi = JoinSplitInterface.abi;
 
+const signNote = (validatorAddress, noteHash, spender, privateKey) => {
```
Should this be in a util file?

thomas-waite replied:
Yep, we have since addressed this on the `develop` branch 

All signing related functions are now in a `signer` module

ArnSch commented on packages/protocol/contracts/ZkAsset/ZkAssetMintableTest.sol:
```diff
@@ -0,0 +1,27 @@
+pragma solidity >=0.5.0 <0.6.0;
```
I'm not a big fan of this pattern, especially not if it means having test contracts in a non-test directory. I understand that you're making sure that a contract can use confidentialTransferFrom, but I don't see what that teaches us compared to say having address[2] call these methods after being approved?

thomas-waite replied:
Yeah good point, that's a nicer way to do it - will update the tests.

We clean away this pattern of having test contracts in non-test directories on `develop`

ArnSch commented on packages/protocol/test/ZkAsset/ZkAssetMintable.js:
```diff
@@ -137,6 +152,208 @@ contract('ZkAssetMintable', (accounts) => {
+                    zkAssetMintableTest.address,
+                    inputNoteOwners[i].privateKey
+                );
+                const concatenatedSignature = signature[0] + signature[1].slice(2) + signature[2].slice(2);
```
This is... ugly. I believe this got changed a lot on `develop` and cleaned away so I think all good.

thomas-waite replied:
I agree. It did get changed to an extent on `develop`, there is a function defined at the top of the `ZkAssetOwnable.js` script: 

    const confidentialApprove = async (indexes, notes, aztecAccounts) => {
        await Promise.all(
            indexes.map((i) => {
                const signature = signer.signNote(
                    zkAssetOwnable.address,
                    notes[i].noteHash,
                    zkAssetOwnableTest.address,
                    aztecAccounts[i].privateKey,
                );
                // eslint-disable-next-line no-await-in-loop
                return zkAssetOwnable.confidentialApprove(notes[i].noteHash, zkAssetOwnableTest.address, true, signature);
            }),
        );
    };

It's a bit nicer, but there's still scope to improve - the function should probably be moved to a `utils` type module, would make it clearer for external users too. Will leave this as is on `master` for now, but make a change on `develop` in my current PR.

ArnSch reviewed (approved):
Some minor points but otherwise LGTM

thomas-waite committed:
test(protocol): remove pattern of using ZkAssetMintableTest contract

can achieve same test effect using a delegateAddress = accounts[2]

thomas-waite committed:
refactor(protocol): move test contract to test directory

thomas-waite committed:
test(protocol): migrate mint confidentialTransferFrom() tests
delete old ZkAssetMintable.js test file and migrate the new tests to the new test file

thomas-waite committed:
docs: doc and eslint fixes

thomas-waite committed:
test(protocol): fix rebase problems
remove canAdjustSupply and canConvert explicit flag declarations, for ZkAssetMintable.js test when  ZkAssets are instantiated
eslint and prettier fixes
